src/content/components/TimelineViewport.js

```diff
@@ -0,0 +1,491 @@
+      }
+      setTimeout(() => {
+        if (scollId === this.shiftScrollId) {
+          this.setState({ isShiftScrollHintVisible: false });
+        }
+      }, 1000);
+    }
+
+    componentDidUpdate(prevProps: Props) {
+      if (this.props.viewportNeedsUpdate(prevProps, this.props)) {
+        this.setState(this.getDefaultState(this.props));
+        requestAnimationFrame(this._setSize);
+      }
+    }
+
+    componentWillReceiveProps(newProps: Props) {
+      if (this.props.isRowExpanded !== newProps.isRowExpanded) {
+        this.setState(this.getDefaultState(newProps));
+        return;
+      }
+      if (
+        this.props.selection !== newProps.selection ||
+        this.props.timeRange !== newProps.timeRange
+      ) {
+        this.setState(this.getHorizontalViewport(newProps));
+      }
+    }
+
+    _setSize() {
+      const rect = this.refs.container.getBoundingClientRect();
+      if (this.state.containerWidth !== rect.width || this.state.containerHeight !== rect.height) {
+        const style = window.getComputedStyle(this.refs.container);
+
+        // Obey margins of the containing element.
+        const containerWidth = rect.width - parseFloat(style.marginLeft) - parseFloat(style.marginRight);
+        const containerHeight = rect.height - parseFloat(style.marginTop) - parseFloat(style.marginBottom);
+        const containerLeft = rect.left + parseFloat(style.marginLeft);
+        const viewportBottom = this.state.viewportTop + containerHeight;
+
+        this.setState({ containerWidth, containerHeight, containerLeft, viewportBottom });
+      }
+    }
+
+    _setSizeNextFrame() {
+      requestAnimationFrame(this._setSize);
+    }
+
+    _mouseWheelListener(event: SyntheticWheelEvent) {
+      if (event.shiftKey) {
+        this.zoomRangeSelection(event);
+        return;
+      }
+
+      if (!this.props.isRowExpanded) {
+        return;
+      }
+
+      // Only move the viewport if the entire canvas is in frame, otherwise let the
+      // TimelineView scrolling element scroll.
+      const maybeScrollElement = this.props.getScrollElement();
+      const viewScrollTop = maybeScrollElement ? maybeScrollElement.scrollTop : 0;
+      const offsetTopAndCollapsedRow = Math.max(0, this.refs.container.offsetTop - COLLAPSED_ROW_HEIGHT);
+
+      // Do not move the contents if the chart is not fully within view.
+      if (event.deltaY < 0) {
+        // Scrolling up.
+        if (viewScrollTop > offsetTopAndCollapsedRow) {
+          return;
+        }
+      } else {
+        if (
+          // The top of the viewport is not within a single collapsed row height.
+          (viewScrollTop < offsetTopAndCollapsedRow) &&
+          // Can the scroll element still scroll?
+          (maybeScrollElement
+            ? maybeScrollElement.scrollTop !== maybeScrollElement.scrollHeight - maybeScrollElement.offsetHeight
+            : false)
+        ) {
+          return;
+        }
+      }
```

Comment:
Can you move this whole block of ifs into a separate method?

And I don't really like the use of the properties scrollTop, scrollHeight and offsetHeight. All of those are rounded CSS pixel values that are an imperfect representation of the actual fractional values that the layout engine uses. The only DOM API that currently uses fractional values is getBoundingClientRect(). Can you do these measurements by using getBoundingClientRect() on inner and outer elements? (That's what I did in the tree view.)
I'm also concerned by the `!==` comparison. It should probably be a `<` or a `>`, so that it still does what you expect in the presence of overscrolling or if there's loss of precision for some other reason.